kapitan/refs/secrets/vault.py

```diff
@@ -0,0 +1,167 @@
+    env['namespace'] = getenv('VAULT_NAMESPACE')
+    # AUTHENTICATION TYPE
+    auth_type = getenv( 'VAULT_AUTHTYPE', default='token')
+    if auth_type == 'token':
+        env['token'] = getenv( 'VAULT_TOKEN' )
+        if not env['token']:
+            with open(join(expanduser('~'),'.vault-token'),'r') as f:
+                env['token'] = f.read()
+    elif auth_type == 'userpass':
+        env['username'] = getenv( 'VAULT_USER' )
+        env['password'] = getenv( 'VAULT_PASSWORD' )
+    elif auth_type == 'approle':
+        env['role_id'] = getenv('VAULT_ROLE_ID')
+        env['secret_id'] = getenv( 'VAULT_SECRET_ID' )
+    # VERIFY VAULT SERVER TLS CERTIFICATE
+    verify = getenv( 'VAULT_SKIP_VERIFY', default='')
+    if verify.lower() == 'true':
+        env['verify'] = False
+    elif verify.lower() == 'false':
+        cert = getenv( 'VAULT_CACERT' )
+        if not cert:
+            cert_path = getenv( 'VAULT_CAPATH' )
+            if not cert_path:
+                raise Exception('Neither VAULT_CACERT nor VAULT_CAPATH specified')
+            env['verify'] = cert_path
+        else:
+            env['verify'] = cert
+    # CLIENT CERTIFICATE FOR TLS AUTHENTICATION
+    client_key,client_cert = getenv( 'VAULT_CLIENT_KEY' ), getenv( 'VAULT_CLIENT_CERT' ) 
+    if client_key != None and client_cert != None:
+        env['cert'] = (client_cert,client_key)
+    return env
+
+def vault_obj():
+    client = hvac.Client(
+        **{k:v for k,v in get_env().items() if v is not None}
+    )
+    assert (
+        client.is_authenticated()
+    ), "Vault Authentication Error, Environment Variables defined?"
+    return client
+
+
+class VaultSecret(Ref):
+
+    """
+    Hashicorp Vault can be used if using KV Secret Engine
+    """
+
+    def __init__(self,data,encode_base64=False,**kwargs):
+        """
+        set encoding_base64 to True to base64 encoding key before encrypting and writing
+        """
+        self._encrypt(data, encode_base64)
+        kwargs['encoding'] = self.encoding
+        super().__init__(self.data,**kwargs)
+        self.type_name = 'vault'
+
+    def _encrypt(self, data,encode_base64):
+        """
+        encrypt data
+        set encode_base64 to True to base64 encode data before writing
+        """
+        self.encoding = "original"
+        if encode_base64:
+            self.data = base64.b64encode(data.encode())
+            self.encoding = "base64"
+        else:
+            self.data = data.encode()
+
+
+    def reveal(self):
+        """
+        returns decrypted data
+        """
+        # can't use super().reveal() as we want bytes
+        ref_data = base64.b64decode(self.data)
+        return self._decrypt(ref_data)
+
+    def _decrypt(self, data):
+        """Decrypt data & return value for the key from Vault Server
+
+        :returns: secret in plain text
+
+        """
+        try:
+            if data.decode() == "secret_test_key":
```

Comment:
Avoid having test case data inside method. Rather use mock python unit test library to achieve the same. Change the coding design to incorporate composition pattern. Refer to https://docs.python.org/3/library/unittest.mock.html